packages/interactivity-router/src/index.ts

```diff
@@ -109,17 +112,17 @@ const regionsToVdom: RegionsToVdom = async ( dom, { vdom } = {} ) => {
 
 // Render all interactive regions contained in the given page.
 const renderRegions = ( page: Page ) => {
-	batch( () => {
+	batch( async () => {
+		populateServerData( page.initialData );
 		if ( globalThis.IS_GUTENBERG_PLUGIN ) {
 			if ( navigationMode === 'fullPage' ) {
 				// Once this code is tested and more mature, the head should be updated for region based navigation as well.
-				updateHead( page.head );
+				await updateHead( page.head );
```

Comment:
Do you think it's possible to move this `await` outside of the `renderRegions()` logic?

I feel `renderRegions()` should be synchronous. 😅 Maybe we can move the `updateHead()` call right before running `renderRegions()`, [here](https://github.com/WordPress/gutenberg/blob/bcb3ff8ea4aaba550529e55e50c0291735133e7e/packages/interactivity-router/src/index.ts#L162) and [here](https://github.com/WordPress/gutenberg/blob/bcb3ff8ea4aaba550529e55e50c0291735133e7e/packages/interactivity-router/src/index.ts#L307).

This way, we wouldn't need an async function inside `renderRegions()`.

Reply:
Good thinking! I will try the other solution - we can preload the scripts in a
separate function like you suggested in the other comment.

Reply:
OK, that does not quite make sense. I've just tried making `renderRegions()` sync but it's not really possible as we need to wait for the assets to be preloaded before updating the HTML.

Reply:
I gave it some thought and tried to make the `renderRegions()` sync but I think
that it's not the right pattern nor is it necessary. 

We always **have to** wait for the new JS to load before rendering the new
HTML. By "new HTML", I mean here HTML of a new page using client-side
full-page navigation or a new region using render-based navigation.

If we don't await the JS that might have been loaded on the new page, we run
into the very same issue that this PR is trying to fix:
https://github.com/WordPress/gutenberg/issues/63880 😃